py/gc.c

```diff
@@ -337,8 +333,18 @@ void gc_collect_start(void) {
 void gc_collect_root(void **ptrs, size_t len) {
     for (size_t i = 0; i < len; i++) {
         void *ptr = ptrs[i];
-        VERIFY_MARK_AND_PUSH(ptr);
-        gc_drain_stack();
+        if (VERIFY_PTR(ptr)) {
+            // Mark and push this pointer
+            // TODO: almost-duplicate code in gc_drain_stack
+            size_t block = BLOCK_FROM_PTR(ptr);
+            if (ATB_GET_KIND(block) == AT_HEAD) {
+                // an unmarked head, mark it, and push it on gc stack
+                TRACE_MARK(block, ptr);
+                ATB_HEAD_TO_MARK(block);
+                *MP_STATE_MEM(gc_sp)++ = block;
+            }
+            gc_drain_stack();
```

Comment:
This can probably move up 1 line into the if-statement.

Reply:
For some reason, moving it up increases code size by 4 bytes in gcc 5.4.1 (for the minimal port).

Reply:
IMO that's acceptable: it's ok to have a tiny increase in size if the code is logically better.  In this case you would expect the code size to be the same if the call is inside or outside the if, and maybe a different gcc (or non-gcc compiler) does something slightly different and makes the code smaller or the same.  And remember that gc.c is actually compiled with -O3 by default (grep for CSUPEROPT) so size optimisations are not always what they seem :)

Is summary, I'd prefer if this line were within the if to make the logic clearer.

Reply:
Done.